### PR TITLE
Additional value forms

### PR DESCRIPTION
**Pull request by milanjaros**
I was not able to find a convenient way to transform `sourceName` to kebab-case nor camelCase even using `replace` with RegExp pattern. 
Additional value form identifiers:
* firstLowerCase
* firstUpperCase
* kebabCase
* titleCase

By a combination of these and `replace`, it will be also possible to convert any `sourceName` to camelCase, PascalCase and snake_case together with kebab-case and Title Case.

Should I also add the `*InvariantValueFormModel`s for each or should I extend their `JObject configuration`s with `bool isInvariant` property?

Just note, I tried hardly to come up with solution, but the best I was able to make kebab-case is following configuration (it fails in some edge cases covered by unit tests). I didn't find any way to do camelCase. Alternative way to resolve camelCase could be to implement `JoinValueForm` for joining strings.

```
  "forms": {
    "kebab": {
      "identifier": "replace",
      "pattern": "([a-zA-Z]{1,4})([A-Z]+)",
      "replacement": "$1-$2"
    },
    "lowerCase": {
      "identifier": "lowerCaseInvariant"
    },
    "kebabCase": {
      "identifier": "chain",
      "steps": [ "kebab", "lowerCase"]
    },
    "alternativeKebab": {
      "identifier": "replace",
      "pattern": "(?<!^)([A-Z][a-z]|(?<=[a-z])[A-Z])",
      "replacement": "-$1"
    }
 }
```

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/templating?pullRequest=2479) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/templating?pullRequest=2479) before we can accept your contribution.<br/><br/>:x: milanjaros [sign now](https://cla.dotnetfoundation.org/dotnet/templating?pullRequest=2479)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/templating?pullRequest=2479) it.</sub>

**Review by grinrag — CHANGES_REQUESTED** [review]
The changes look useful to me. But could, you, please merge your commits into one (especially the last two ones)?

**Comment by milanjaros** [comment]
@grinrag, not sure what do you mean, should I create a new pull-request with squashed commits?

And please, should I also add the `*InvariantValueFormModel`s for each?

**Comment by grinrag** [comment]
@milanjaros it's not necessary to create a new PR, you can just update your branch (with `--force`, since you squashed the commits).
Yes, let's do the `InvariantValueFormModel` for each new model since we have them for old ones.

What is the problem with kebab-case and what edge case, you mentioned above?

**Comment by grinrag** [comment]
@milanjaros Hello Milan. Do you have some additional questions or are you planning to update the PR? Thanks!

**Comment by vlada-shubina** [comment]
@milanjaros we also have documentation for value forms in wiki: https://github.com/dotnet/templating/blob/master/docs/Runnable-Project-Templates---Value-Forms.md - do you mind please add the forms you've added to the doc? 

Thanks a lot.

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
Please add short description of the forms to the documentation.

**Comment by vlada-shubina** [comment]
@milanjaros I've done some tests on this change: e2e didn't work due to forms were not set up in ValueFormModel.cs, so fixed that. I've also added documentation. From my side it ready to be merged now.

**Review by vlada-shubina — APPROVED** [review]

**Review by AR-May — APPROVED** [review]
LGTM

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstLowerCaseValueFormModel.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            return new FirstLowerCaseValueFormModel(name);
+        }
+
+        public string Process(IReadOnlyDictionary<string, IValueForm> forms, string value)
+        {
+            switch (value)
+            {
+                case null: return null;
+                case "": return value;
+                default: return value.First().ToString().ToLower() + value.Substring(1);
```
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstLowerCaseValueFormModel.cs:34]: `ToLower` is culture-dependent. Is this what we want or should we change it to use `CultureInfo.InvariantCulture`?

Comment (bozturkMSFT) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstLowerCaseValueFormModel.cs:34]: Same goes for `ToUpper` below.

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstLowerCaseValueFormTests.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+namespace Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests.ValueFormTests
+{
+    public class FirstLowerCaseValueFormTests
+    {
+        [Theory]
+        [InlineData("A", "a")]
+        [InlineData("NO", "nO")]
+        [InlineData("NEW", "nEW")]
+        [InlineData("", "")]
+        [InlineData(null, null)]
```
Comment (bozturkMSFT) [inline test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstLowerCaseValueFormTests.cs:13]: Regardless of what culture related decision we make, I think we should add tests for different cultures.

**Review by bozturkMSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstLowerCaseValueFormModel.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstLowerCaseValueFormModel.cs:34]: There was a question whether we need invariant versions above. We do have lowerCase and lowerCaseInvariant forms.

To be honest I'm not quite sure if we should do two versions here. I believe that this form is needed in invariant culture, as I see the use case to generate variables names, which are usually invariant.  However, keeping firstLowerCase as Invariant and lowerCase as culture specific might be confusing. 

Given said that, I would rather do 2 forms (firstLowerCase and firstLowerCaseInvariant) and pass the responsibility to template author which one to choose (and if it is used correctly in given context :) ).

Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstLowerCaseValueFormModel.cs:34]: Implemented invariant forms in new commit and updated test.

**Review by bozturkMSFT — APPROVED** [review]

**Comment by vlada-shubina** [comment]
I have also updated JSON schema, as it contains definition for value forms. Last minute update for kebabCase form to use ToLowerInvariant. Updated wiki doc to mention version compatibility.

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/KebabCaseValueFormTests.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        [InlineData("WebAPI", "web-api")]
+        [InlineData("XProjectX", "x-project-x")]
+        [InlineData("NextXXXProject", "next-xxx-project")]
+        [InlineData("NoNewProject", "no-new-project")]
+        [InlineData("NONewProject", "no-new-project")]
+        [InlineData("NewProjectName", "new-project-name")]
+        [InlineData("ABBREVIATIONAndSomeName", "abbreviation-and-some-name")]
+        [InlineData("NoNoNoNoNoNoNoName", "no-no-no-no-no-no-no-name")]
+        [InlineData("AnotherNewNewNewNewNewProjectName", "another-new-new-new-new-new-project-name")]
+        [InlineData("Param1TestValue", "param-1-test-value")]
```
Comment (AR-May) [inline test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/KebabCaseValueFormTests.cs:24]: Consider including the case when we have 2 or more consecutive digits in the string. (https://stackoverflow.com/a/54012346/164680 gives a good set of use cases, I suppose).

**Review by AR-May — CHANGES_REQUESTED** [review]

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/KebabCaseValueFormModel.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        public IValueForm FromJObject(string name, JObject configuration)
+        {
+            return new KebabCaseValueFormModel(name);
+        }
+
+        /// <summary>
+        /// PascalCase to kebab-case using Microsoft's capitalization conventions (https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions).
+        /// Todd Skelton's solution (https://stackoverflow.com/a/54012346/164680)
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/KebabCaseValueFormModel.cs:29]: Actually the solution at stack overflow was edited 5 days ago (https://stackoverflow.com/posts/54012346/revisions), that's way it is not included.

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstUpperCaseInvariantValueFormModel.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.TemplateEngine.Orchestrator.RunnableProjects.ValueForms
+{
+    public class FirstUpperCaseInvariantValueFormModel : IValueForm
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/FirstUpperCaseInvariantValueFormModel.cs:8]: All the value forms classes are not sealed at the moment. I would do this change with other refactoring on this topic.

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/KebabCaseValueFormModel.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            if (value is null)
+            {
+                return null;
+            }
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                return string.Empty;
+            }
+
+            Regex pattern = new Regex(@"(?:\p{Lu}\p{M}*)?(?:\p{Ll}\p{M}*)+|(?:\p{Lu}\p{M}*)+(?!\p{Ll})|\p{N}+|[^\p{C}\p{P}\p{Z}]+|[\u2700-\u27BF]");
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ValueForms/KebabCaseValueFormModel.cs:40]: Unicode wise pattern to determine words in the string

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseValueFormTests.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        [InlineData("new", "New", null)]
+        [InlineData("", "", null)]
+        [InlineData(null, null, null)]
+        [InlineData("indigo", "İndigo", "tr-TR")]
+        [InlineData("ındigo", "Indigo", "tr-TR")]
+        public void FirstUpperCaseWorksAsExpected(string input, string expected, string culture)
+        {
+            if (!string.IsNullOrEmpty(culture))
+            {
+                if (culture == "invariant")
```
Comment (DavidKarlas) [inline test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseValueFormTests.cs:21]: Is this meant to be `null`?

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseInvariantValueFormModel.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        [InlineData("new", "New", null)]
+        [InlineData("", "", null)]
+        [InlineData(null, null, null)]
+        [InlineData("indigo", "Indigo", "tr-TR")]
+        [InlineData("ındigo", "ındigo", "tr-TR")]
+        public void FirstUpperCaseInvariantWorksAsExpected(string input, string expected, string culture)
+        {
+            if (!string.IsNullOrEmpty(culture))
+            {
+                if (culture == "invariant")
```
Comment (DavidKarlas) [inline test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseInvariantValueFormModel.cs:21]: Is this meant to be `null`?

**Review by DavidKarlas — APPROVED** [review]
Take a look at `culture = null` vs `culture == "invariant"` otherwise looks OK

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseValueFormTests.cs (hunk 7 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseValueFormTests.cs:21]: No, it is not. 
This snippet sets the culture if not null in data. 
To use invariant culture, `"invariant"` const should be used. It is not used in this test in particular, but other tests are using it.
If data is null, current culture is used. 

The only thing came to my mind, is that we should probably reset it once test is done, otherwise other tests will be executed with last culture set.

Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/ValueFormTests/FirstUpperCaseValueFormTests.cs:21]: I would rather address it in another PR, as there are other tests which do not reset culture.

**Review by AR-May — APPROVED** [review]

**Comment by vlada-shubina** [comment]
Thank you @milanjaros - it is merged now.
